### ISSUE TITLE
Issue with inferencing on a tensorrt model (converted from yolo_nas_s)

### ISSUE THREAD

**Issue by Atm-ninetyseven** [Deci-AI/super-gradients#1200]
### 💡 Your Question

I am experiencing difficulties when attempting to perform inference on a TensorRT model. The model was generated through the following procedure:

- Yolo nas s pytorch model and quantized model created by following the given 'Quantization Aware Training YoloNAS on Custom Dataset Notebook' google colab, which creates two onnx models (QAT and PQT). 
- Tensorrt model created with the following command:
        - ```trtexec --onnx=<qat onnx model save path> --saveEngine=<trt_model_save_path> --explicitBatch=4 --fp16 --int8```

This process gives me a trt model, which is then used with the following code:

### Loading the model:
```
import tensorrt as trt
engine_path = '<trt model save path>'
logger = trt.Logger(trt.Logger.WARNING)

runtime = trt.Runtime(logger)
trt.init_libnvinfer_plugins(logger,'')

with open(engine_path, "rb") as f:
    serialized_engine = f.read()

engine = runtime.deserialize_cuda_engine(serialized_engine)
context = engine.create_execution_context()

imgsz = engine.get_binding_shape(0)[2:]
context = engine.create_execution_context()

inputs, outputs, bindings = [], [], []
stream = cuda.Stream()
for binding in engine:
    size = trt.volume(engine.get_binding_shape(binding))
    dtype = trt.nptype(engine.get_binding_dtype(binding))
    host_mem = cuda.pagelocked_empty(size, dtype)
    device_mem = cuda.mem_alloc(host_mem.nbytes)
    bindings.append(int(device_mem))
    if engine.binding_is_input(binding):
        inputs.append({'host': host_mem, 'device': device_mem})
    else:
        outputs.append({'host': host_mem, 'device': device_mem})
```
### Loading and preprocessing the image:
```
def preprocess(image):
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    mean = np.array([0.485, 0.456, 0.406]).astype('float32')
    stddev = np.array([0.229, 0.224, 0.225]).astype('float32')
    data = (np.asarray(image).astype('float32') / float(255.0) - mean) / stddev

    return np.moveaxis(data, 2, 0)

#to keep things simple:
image_path =<image path>
image_cv2 = cv2.imread(image_path)
image_cv2 = cv2.resize(image_cv2, (640, 640))
images = [image_cv2, image_cv2, image_cv2, image_cv2]

preprocessed_images = []
for image in images:
    preprocessed_image = preprocess(image=image.copy())
    preprocessed_images.append(preprocessed_image)
```

### Inference code:
```
import torch

preprocessed_images_array = np.array(preprocessed_images).astype(np.float32)
inputs[0]['host'] = preprocessed_images_array

for inp in inputs:
    cuda.memcpy_htod_async(inp['device'], inp['host'], stream)
context.execute_async_v2(bindings=bindings, stream_handle=stream.handle)
for out in outputs:
    cuda.memcpy_dtoh_async(out['host'], out['device'],stream)
stream.synchronize()

data = [out['host'] for out in outputs]
```

### Postprocessing:
```
boxes = torch.from_numpy(data[1]).view(4,8400,4)
scores = torch.from_numpy(data[0]).view(4,8400,4)

model_output = ((boxes, scores),()) #so that i can feed it through the post prediction callback

post_nms_predictions = post_prediction_callback(model_output, device=self.device)[0].numpy() 
#where post_prediction_callback is 'super_gradients.training.models.detection_models.pp_yolo_e.post_prediction_callback.PPYoloEPostPredictionCallback'
```
The results seem to be terrible. Is the above pipeline accurate? If not, what could be the potential source of error? Are the preprocessing and postprocessing same for the pytorch model and tensorrt model? 

Notes on the pytorch model:
- The pytorch model has the following preprocessing steps: reverse image channels, rescale and padding, standardizing to 0-1 (with max value 255, but not considering mean and std), and image permute (2, 0, 1). Note that the standardizing step of the pytorch model is different to what I perform with the tensorrt, as when the image was standardized without considering the mean and std, the tensorrt output boxes are on the edge. 

### Versions

PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N[/A](https://file+.vscode-resource.vscode-cdn.net/A)

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 8.4.0-3ubuntu2) 8.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-73-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2070
Nvidia driver version: 530.30.02
cuDNN version: Probably one of the following:
[/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5](https://file+.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5)
[/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.2](https://file+.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.2)
[/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.2](https://file+.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.2)
[/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.2](https://file+.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.2)
[/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.2](https://file+.vscode-resource.vscode-cdn.net/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.2)
...
[conda] torchaudio                0.11.0+cu113             pypi_0    pypi
[conda] torchinfo                 1.8.0                    pypi_0    pypi
[conda] torchmetrics              0.8.0                    pypi_0    pypi
[conda] torchvision               0.12.0+cu113             pypi_0    pypi

**Comment by ChuRuaNh0** [comment]
https://github.com/Deci-AI/super-gradients/issues/1288

**Comment by harpreetsahota204** [comment]
Hey @Atm-ninetyseven @ChuRuaNh0 

Thanks for opening an issue for SG. I'm gathering some feedback on SuperGradients and YOLO-NAS.

Would you be down for a quick call to chat about your experience?

If a call doesn't work for you, no worries. I've got a short survey you could fill out: https://bit.ly/sgyn-feedback.

I know you’re super busy, but your input will help us shape the direction of SuperGradients and make it as useful as possible for you.

I appreciate your time and feedback. Let me know what works for you.

Cheers,

Harpreet
